### PR TITLE
Unsubcribe email

### PR DESCRIPTION
**Pull request by nicosantangelo**
# Important

It *does not* enforce the modal on first view to keep the intro modal flow. This is easily changed but we may have to discuss it further.

For example another option is to open the modal after the first successful bid

We also need a copy for the unsubscribe page:

image

image

### CODE REVIEW

**Comment by abarmat** [comment]
Link your account works great!

1) There's a small typo: recieve => receive
2) What do you think if we make the checkbox clickable? I instinctively tried to click a button to close the modal
3) I also managed to insert an email of the format ab@g without a proper domain name

link-account

File: webapp/src/lib/api.js [hunk 1]
```diff
@@ -30,15 +30,27 @@ class API {
...
-  deleteOutbidNotification(email) {
-    return this.request('delete', '/outbidNotification', { email })
+  postOutbidNotification(address, parcelStateIds) {
+    return this.request('post', '/outbidNotification', {
+      address,
+      parcelStateIds: parcelStateIds.join(';')
+    })
+  }
+
+  deleteOutbidNotifications(address) {
```
Comment (abarmat) [inline webapp/src/lib/api.js]: What if we remove this endpoint which allows anyone to unsubscribe you and send the signed message when we click on the unsubscribe button in the sidebar?

**Comment by abarmat** [comment]
Instead of getting the email address from the state what if we use a param which is easy to link from the email and works even if the user lost the current state => /unsubscribe/abarmat@gmail.com

Then when we send the signature we do message => address => email === url.email ?

What do you think?

File: webapp/src/lib/api.js (hunk 1 above)
Comment (nicosantangelo) [reply webapp/src/lib/api.js]: Oh you're right!

We still have this problem when you add new bids. We do:

```javascript
function* handleEmailRegisterBids(action) {
  const { address } = yield select(selectors.getAddressStateData)

  if (address) {
    const parcelStateIds = action.bids.map(bid => buildCoordinate(bid.x, bid.y))
    yield call(() => api.postOutbidNotification(address, parcelStateIds))
  }
}
```

and I don't think we're checking the user actually has bids on the parcels.
Someone else could send that request and subscribe someone else to some bids (only if he/she already has an email registered)

Should we fix that too?

**Comment by nicosantangelo** [comment]
> Instead of getting the email address from the state what if we use a param which is easy to link from the email and works even if the user lost the current state => /unsubscribe/abarmat@gmail.com

I actually implemented it like that first, but because the user needs to use metamask to sign the unsubscription I thought it doesn't make sense to have a query string parameter that can be meddled with.
The current `handleEmailUnsubscribeNewsletter` only uses information that gets from the server, that in turn gets from MetaMask so I think it's a bit more secure no?

**Comment by nicosantangelo** [comment]
> There's a small typo: recieve => receive

fixing

> What do you think if we make the checkbox clickable? I instinctively tried to click a button to close the modal

Yes!

> I also managed to insert an email of the format ab@g without a proper domain name

Hm you think we need to validate for mainstream providers? I didn't thought of that because we already have the confirmation email (don't we? :P)

**Comment by nicosantangelo** [comment]
Updated!

**Comment by abarmat** [comment]
> I also managed to insert an email of the format ab@g without a proper domain name
> 
> Hm you think we need to validate for mainstream providers? I didn't thought of that because we already have the confirmation email (don't we? :P)
> 
I think it will be rejected by the email provider as an invalid address

**Comment by abarmat** [comment]
> Instead of getting the email address from the state what if we use a param which is easy to link from the email and works even if the user lost the current state => /unsubscribe/abarmat@gmail.com
> 
> I actually implemented it like that first, but because the user needs to use metamask to sign the unsubscription I thought it doesn't make sense to have a query string parameter that can be meddled with.
> The current handleEmailUnsubscribeNewsletter only uses information that gets from the server, that in turn gets from MetaMask so I think it's a bit more secure no?

~~What I tried to mean with the comment is that if the user clicks on the unsubscribe email and gets to the app, they lost their state but they have Metamask, they won't be able to unsubscribe. Is that correct?~~

All good :)

File: webapp/src/lib/api.js (hunk 1 above)
Comment (abarmat) [reply webapp/src/lib/api.js]: Yes, you are right, saw that too! but it's more difficult as we have to sign the message again when adding new bids to validate you are the one registering the parcels. Maybe keeping a signature in localStorage?

For not allowing registration on parcels you haven't bid we could add that validation I think on the registerParcelNotifications() method. Right?

**Comment by nicosantangelo** [comment]
Updated!

**Comment by abarmat** [comment]
It's ok, good to go!

**Comment by eordano** [comment]
Yep!

**Pull request merged**